Implement Protobuf support

amrutha-shanbhag opened the pull request:
<!-- All contributors please complete these sections, including maintainers -->
# About this change - What it does


- Validating, storing, and versioning Protobuf schemas
- Protobuf schema evolution and compatibility
- Protobuf messages serialization/deserialization 

<!-- Provide the issue number below if it exists. -->
References: #67 

# Why this way

Design document outlining design decisions: https://github.com/instaclustr/karapace/pull/11/files?short_path=45bec23#diff-45bec23eda89f362ca9cda50a6be0477f6058d4e6b9cc6c5198cdc5f0488df77 
<!-- Provide a small explanation on why this is the approach you took for solving this problem. -->

amrutha-shanbhag commented:
> Thanks for this contribution @amrutha-shanbhag! There's a lot of code here to review, but I see that some people have already had discussions in the issue #67 and your design document. We'll review your PR as soon as possible.

Hi @RyanSkraba , you are right. The code was developed incrementally on Instaclustr fork, and regularly reviewed by @hackaugusto and @juha-aiven 

Thanks

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+
+    def __init__(self, writer_schema=None):
+        self._writer_schema = writer_schema
+        a: ProtobufSchema = writer_schema
```
This is unnecessary, because the constructor argument shouldn't be optional.

libretto replied:
optional constructor arguments come from template See. avro/io.py

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+        self._writer_schema = writer_schema
+        a: ProtobufSchema = writer_schema
+        el: TypeElement
+        self._message_name = ''
```
```suggestion
        self._message_name = ""
```

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+                self._message_index = idx
+                break
+
+        if self._message_name == '':
```
```suggestion
        if self._message_name == "":
```

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+        if self._message_name == '':
+            raise ProtobufTypeException("No message in protobuf schema")
+
+    def write_index(self, writer: BytesIO):
```
```suggestion
    def write_index(self, writer: BytesIO) -> None:
```

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+    def write_index(self, writer: BytesIO):
+        write_indexes(writer, [self._message_index])
+
+    def write(self, datum: dict, writer: BytesIO):
```
```suggestion
    def write(self, datum: dict, writer: BytesIO) -> None:
```

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+class ProtobufDatumReader:
+    """Deserialize Protobuf-encoded data into a Python data structure."""
+
+    def __init__(self, writer_schema=None, reader_schema=None):
```
Please make both args required, their types are `ProtobufSchema`:

```suggestion
    def __init__(self, writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> None:
```

hackaugusto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+        self._reader_schema = reader_schema
+
+    def read(self, bio: BytesIO):
+        if self._reader_schema is None:
```
this can go away once the constructors args are no longer optional

hackaugusto commented on karapace/protobuf/kotlin_wrapper.py:
```diff
@@ -0,0 +1,59 @@
+        raise IllegalStateException(message)
+
+
+def trim_margin(s: str) -> str:
```
please replace this with https://docs.python.org/3/library/textwrap.html?highlight=dedent#textwrap.dedent

libretto replied:
This function contains a bit more logic than textwrap.dedent() which actually is used inide. To replace it with dedent() I need to change the format of test data in unit tests. But it is a lot of work plus the format of data in these tests come from square/wire project and has the sense.

hackaugusto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+        """
+        return None if self.is_scalar or self.is_map else f"type.googleapis.com/{self.string}"
+
+    def nested_type(self, name: str) -> object:  # ProtoType
```
```suggestion
    def nested_type(self, name: str) -> "ProtoType":
```

hackaugusto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+                # TODO: must be IllegalArgumentException
+                raise Exception(f"map key must be non-byte, non-floating point scalar: {key_type}")
+
+    def to_kind(self) -> OptionElement.Kind:
```
```suggestion
    def to_kind(self) -> Optional[OptionElement.Kind]:
```

hackaugusto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+
+    @property
+    def enclosing_type_or_package(self) -> str:
+        """ Returns the enclosing type, or null if self type is not nested in another type.  """
```
```suggestion
        """Returns the enclosing type, or None if self type is not nested in another type."""
```

hackaugusto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+        return None if (dot == -1) else self.string[:dot]
+
+    @property
+    def type_url(self) -> str:
```
```suggestion
    def type_url(self) -> Optional[str]:
```

hackaugusto commented on karapace/protobuf/protobuf_to_dict.py:
```diff
@@ -0,0 +1,336 @@
+# -*- coding:utf-8 -*-
```
this module is missing typing information.

libretto replied:
it is 3rd party module which with minor fixes to be compliant to our project.

hackaugusto commented on karapace/protobuf/reserved_document.py:
```diff
@@ -0,0 +1,35 @@
+from karapace.protobuf.utils import append_documentation
+
+
+class ReservedElement:
```
This should be a dataclass

libretto replied:
ok

hackaugusto commented on karapace/protobuf/rpc_element.py:
```diff
@@ -0,0 +1,48 @@
+from karapace.protobuf.utils import append_documentation, append_indented
+
+
+class RpcElement:
```
This should be a dataclass

libretto replied:
ok

hackaugusto commented on karapace/protobuf/service_element.py:
```diff
@@ -0,0 +1,40 @@
+from typing import List
+
+
+class ServiceElement:
```
this should be a dataclass

libretto replied:
ok

hackaugusto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,367 @@
+                radix = 16
+            return int(tag, radix)
+
+        #        except OSError as err:
```
dead code

libretto replied:
ok

hackaugusto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,367 @@
+        #            print("OS error: {0}".format(err))
+        except ValueError:
+            self.unexpected(f"expected an integer but was {tag}")
+        return -22  # this return never be called but mypy think we need it
```
this happens because mypy doesn't know that `unexpected` never returns, the type of that method should be fixed and this line removed

libretto replied:
ok

hackaugusto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,367 @@
+        if not condition:
+            self.unexpected(message, location)
+
+    def unexpected(self, message: str, location: Location = None) -> None:
```
https://mypy.readthedocs.io/en/stable/more_types.html#the-noreturn-type

```suggestion
    def unexpected(self, message: str, location: Location = None) -> NoReturn:
```

hackaugusto commented on karapace/protobuf/utils.py:
```diff
@@ -0,0 +1,70 @@
+from typing import List
+
+
+def protobuf_encode(a: str) -> str:
```
deadcode

hackaugusto commented on karapace/schema_reader.py:
```diff
@@ -38,6 +43,17 @@ def parse_jsonschema_definition(schema_definition: str) -> Draft7Validator:
     return Draft7Validator(schema)
 
 
+def parse_protobuf_schema_definition(schema_definition: str) -> ProtobufSchema:
```
This function can be removed and `ProtobufSchema` can be called directly

libretto replied:
Yes but I prefer to keep it in this view to have the similar code looks.

hackaugusto commented on karapace/schema_reader.py:
```diff
@@ -71,25 +87,44 @@ def parse_avro(schema_str: str):  # pylint: disable=inconsistent-return-statemen
             raise InvalidSchema from e
 
+    @staticmethod
+    def parse_protobuf(schema_str: str):
```
```suggestion
    def parse_protobuf(schema_str: str) -> Optional[TypedSchema]:
```

hackaugusto commented on karapace/schema_reader.py:
```diff
@@ -71,25 +87,44 @@ def parse_avro(schema_str: str):  # pylint: disable=inconsistent-return-statemen
+            TypeError, SchemaError, AssertionError, ProtobufParserRuntimeException, IllegalStateException,
+            IllegalArgumentException, ProtobufError, ProtobufException, ProtobufSchemaParseException
+        ) as e:
+            log.exception("Unexpected error: %s \n schema:[%s]", e, schema_str)
```
Question, is this the correct behavior? Shouldn't the exception be propagated? If not, this is missing an explicit `return None`

libretto replied:
yes we

hackaugusto commented on karapace/schema_registry_apis.py:
```diff
@@ -276,6 +276,7 @@ def send_delete_subject_message(self, subject, version):
         value = '{{"subject":"{}","version":{}}}'.format(subject, version)
         return self.send_kafka_message(key, value)
 
+    # TODO: PROTOBUF add protobuf compatibility_check
```
this is implemented, right?

libretto replied:
yes

hackaugusto commented on karapace/serialization.py:
```diff
@@ -134,6 +140,7 @@ def get_subject_name(self, topic_name: str, schema: str, subject_type: str, sche
             namespace = schema_typed.schema.namespace
         if schema_type is SchemaType.JSONSCHEMA:
             namespace = schema_typed.to_json().get("namespace", "dummy")
+        # TODO: PROTOBUF* Seems protobuf does not use namespaces in terms of AVRO
```
what is the status of this todo?

libretto replied:
edited

hackaugusto commented on pytest.ini:
```diff
@@ -1,3 +1,3 @@
 [pytest]
-addopts = -ra -q --tb=short --showlocals --numprocesses auto
+addopts = -ra  --tb=short --showlocals --numprocesses auto
```
please undo this change

hackaugusto reviewed (changes requested):
mypy should be enabled for karapace/protobuf

hackaugusto commented:
@amrutha-shanbhag sorry for the delay in the review, this was on my plate but I took a few days off during the holiday season. I did a first pass, thanks for the contribution :)

libretto commented on karapace/compatibility/__init__.py:
```diff
@@ -63,6 +64,10 @@ def check_jsonschema_compatibility(reader: Draft7Validator, writer: Draft7Valida
     return jsonschema_compatibility(reader, writer)
 
 
+def check_protobuf_compatibility(reader, writer) -> SchemaCompatibilityResult:
```
I try follow to karapace code style. There is the same way as in check_jsonschema_compatibility() ...

libretto commented on karapace/compatibility/protobuf/checks.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def check_protobuf_schema_compatibility(reader: ProtobufSchema, writer: ProtobufSchema) -> SchemaCompatibilityResult:
+    result: CompareResult = CompareResult()
```
SchemaCompatibilityResult is result type of upper level  function check_compatibility() which aggregate compatibility check for all formats. But compatibility of protobuf schemas need internal class CompareResult which used to collect modification, temporary data, and find compatibility.

libretto commented on karapace/config.py:
```diff
@@ -52,7 +52,8 @@
     "karapace_registry": False,
-    "master_election_strategy": "lowest"
+    "master_election_strategy": "lowest",
+    "protobuf_runtime_directory": "runtime"
```
ok

libretto commented on karapace/kafka_rest_apis/__init__.py:
```diff
@@ -25,9 +25,10 @@
+KNOWN_FORMATS = {"json", "avro", "protobuf", "binary"}
 OFFSET_RESET_STRATEGIES = {"latest", "earliest"}
-SCHEMA_MAPPINGS = {"avro": SchemaType.AVRO, "jsonschema": SchemaType.JSONSCHEMA}
+# TODO: PROTOBUF* check schema mapping
```
it done

libretto commented on karapace/protobuf/compare_result.py:
```diff
@@ -0,0 +1,79 @@
+
+
+class Modification(Enum):
+    # TODO
```
so i suppose i add there all possible changes but probably some more logic maybe added later

libretto commented on karapace/protobuf/compare_result.py:
```diff
@@ -0,0 +1,79 @@
+
+class CompareResult:
+    def __init__(self) -> None:
+        self.result: list = []
```
ok

libretto commented on karapace/protobuf/compare_result.py:
```diff
@@ -0,0 +1,79 @@
+        self.path: list = []
+        self.canonical_name: list = []
+
+    def push_path(self, string: str, canonical: bool = False) -> None:
```
it represent part of protobuf message field name (path to field)

libretto commented on karapace/protobuf/compare_result.py:
```diff
@@ -0,0 +1,79 @@
+    def push_path(self, string: str, canonical: bool = False) -> None:
+        if canonical:
+            self.canonical_name.append(str(string))
+        self.path.append(str(string))
```
will fix it

libretto commented on karapace/protobuf/compare_type_storage.py:
```diff
@@ -0,0 +1,130 @@
+        self.other_package_name = other_package_name
+        self.self_types: Dict[str, Union[TypeRecord, TypeRecordMap]] = {}
+        self.other_types: Dict[str, Union[TypeRecord, TypeRecordMap]] = {}
+        self.locked_messages: List[object] = []
```
sorry but most places where You see 'object' is because of circular dependencies

hackaugusto replied:
For typing purposes the solution is to use the constant `TYPE_CHECKING`, here is additional documentation:

https://mypy.readthedocs.io/en/stable/runtime_troubles.html?highlight=circular#import-cycles

libretto replied:
i will try implement it

libretto commented on karapace/protobuf/enum_element.py:
```diff
@@ -0,0 +1,61 @@
+        result.append("}\n")
+        return "".join(result)
+
+    def compare(self, other: 'EnumElement', result: CompareResult, types: CompareTypes) -> None:
```
we have one instance of result because it contains common data for all compare functions in the chain. This instance we create before comparison of two schemas and at final it contains a data which required for last calculation of compatibility.

libretto commented on karapace/protobuf/exception.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class ProtobufSchemaResolutionException(ProtobufException):
+    def __init__(self, fail_msg: str, writer_schema=None, reader_schema=None) -> None:
+        writer_dump = pretty_print_json(str(writer_schema))
+        reader_dump = pretty_print_json(str(reader_schema))
```
Not agree. As template for logic of ProtobufDatumReader class I used DatumReader from  avro/io.py where 
we see `def __init__(self, writer_schema=None, reader_schema=None):`

libretto commented on karapace/protobuf/field.py:
```diff
@@ -0,0 +1,13 @@
+# Ported from square/wire:
+# wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Field.kt
+
+# TODO: We ported only code which is used by schema parser. Full class may be ported if required
```
ok

libretto commented on karapace/protobuf/compare_type_storage.py:
```diff
@@ -0,0 +1,130 @@
+            return name[(len(type_record.package_name) + 1):]
+        return name
+
+    def lock_message(self, message: object) -> bool:
```
circular dependency...

libretto commented on karapace/protobuf/compare_type_storage.py:
```diff
@@ -0,0 +1,130 @@
+        self.locked_messages.append(message)
+        return True
+
+    def unlock_message(self, message: object) -> bool:
```
circular dependency

libretto commented on karapace/protobuf/compare_type_storage.py:
```diff
@@ -0,0 +1,130 @@
+        return False
+
+
+class TypeRecord:
+    def __init__(self, package_name: str, type_element: TypeElement) -> None:
+        self.package_name = package_name
+        self.type_element = type_element
```
ok

libretto commented on karapace/protobuf/compare_type_storage.py:
```diff
@@ -0,0 +1,130 @@
+        try:
+            self.key = key
+            self.value = value
+        except Exception:
```
ok

libretto commented on karapace/protobuf/enum_element.py:
```diff
@@ -0,0 +1,61 @@
+
+class EnumElement(TypeElement):
+    def __init__(
+        self, location: Location, name: str, documentation: str = "", options: list = None, constants: list = None
```
1) Arguments with defaults cannot be declared before arguments without defaults. And Your offer requires to reorder of arguments in this call. 

2) This class is ported from square/wire and I prefer to keep the original interface with order of arguments with similar defaults. 
So I offer to use following construction:
` def __init__(
        self, location: Location,
        name: str,
        documentation: str = "",
        options: List[OptionElement] = None,
        constants: List[EnumConstantElement] = None
    ) -> None:`

libretto commented on karapace/protobuf/enum_element.py:
```diff
@@ -0,0 +1,61 @@
+from karapace.protobuf.utils import append_documentation, append_indented
+
+
+class EnumElement(TypeElement):
+    def __init__(
+        self, location: Location, name: str, documentation: str = "", options: list = None, constants: list = None
+    ) -> None:
+        # Enums do not allow nested type declarations.
+        super().__init__(location, name, documentation, options or [], [])
+        self.constants = constants or []
```
This class and class hierarchy was ported from square/wire. I prefer to keep this class and hierarchy unchanged. 
And because @dataclass is not so flexible I see no acceptable way to use @dataclass construction for class EnumElement without breaking of the class hierarchy.

libretto commented on karapace/protobuf/group_element.py:
```diff
@@ -0,0 +1,40 @@
+        result: list = []
+        append_documentation(result, self.documentation)
+
+        # TODO: compare lower() to lowercase() and toLowerCase(Locale.US) Kotlin
```
It is still TODO. But simple tests show they work in similar ways.

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def match_schemas(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> bool:
+    # TODO (serge): schema comparison by fields required
```
It is my opinion. I suppose that we could use schema comparison, not by dumps but field by filed comparison... But I do not have enough use cases for this code to understand if it is really needed.

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+    return "c_" + hashlib.md5(name.encode('utf-8')).hexdigest()
+
+
+def check_props(schema_one, schema_two, prop_list):
```
ok

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+    try:
+        size: int = read_varint(bio)
+    except EOFError:
+        # TODO: change exception
```
I have a doubt - must I replace this EOFError usage there with some more specific or not.

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def read_data(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema, bio: BytesIO):
+    # TODO (serge): check and polish it
```
Code works and tests are passed. But this code is still in beta and needs to be tested in a real application.

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+class ProtobufDatumWriter:
+    """ProtobufDatumWriter for generic python objects."""
+
+    def __init__(self, writer_schema=None):
```
Much code which I write there I write by the template to follow karapace code style. 
This code is written by avrop/io.py as a template.  So writer_schema=None comes from there.

libretto commented on karapace/protobuf/message_element.py:
```diff
@@ -0,0 +1,144 @@
+from typing import List
+
+
+class MessageElement(TypeElement):
```
@dataclass has limitations on subclass constructor defaults. I prefer to keep the interfaces of this class in sync with sqare/wire project.

libretto commented on karapace/protobuf/message_element.py:
```diff
@@ -0,0 +1,144 @@
+
+                result.pop_path()
+
+            # TODO Compare NestedTypes must be there.
```
This code pass SR unit tests (except tests with dependencies that we do not support yet). This comparison must be implemented together with the implementation of dependencies support for protobuf schemas.

libretto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+        return self.string[dot + 1:]
+
+    @classmethod
+    def static_init(cls) -> None:
```
This code was ported from Kotlin so I packed it in the same class with python native tricks. 
At least my mypy and pylint still accept it without any messages.

libretto replied:
Btw when mypy targed will be added in karapace Makefile?

libretto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+                self.is_scalar = False
+                self.string = string
+                self.is_map = True
+                self.key_type = key_type  # TODO restrict what's allowed here
```
it comes from square/wire project. I saw no changes in their code yet so I think we can get rid of this todo yet.

libretto commented on karapace/protobuf/option_reader.py:
```diff
@@ -0,0 +1,153 @@
+from typing import Union
+
+
+class KindAndValue:
```
done

libretto commented on karapace/protobuf/enum_constant_element.py:
```diff
@@ -0,0 +1,31 @@
+from karapace.protobuf.utils import append_documentation, append_options
+
+
+class EnumConstantElement:
```
done

libretto commented on karapace/protobuf/extend_element.py:
```diff
@@ -0,0 +1,25 @@
+from karapace.protobuf.utils import append_documentation, append_indented
+
+
+class ExtendElement:
```
done

libretto commented on karapace/protobuf/proto_type.py:
```diff
@@ -0,0 +1,213 @@
+        return self.string
+
+    def hash_code(self) -> int:
+        return hash(self.string)
```
What do You mean by 'very odd'?

libretto commented on karapace/protobuf/protobuf_syntax.py:
```diff
@@ -0,0 +1,5 @@
+from enum import Enum
```
ok

libretto commented on karapace/protobuf/protobuf_to_dict.py:
```diff
@@ -0,0 +1,336 @@
+# -*- coding:utf-8 -*-
```
ok

libretto commented on karapace/protobuf/extensions_element.py:
```diff
@@ -0,0 +1,39 @@
+from karapace.protobuf.utils import append_documentation, MAX_TAG_VALUE
+
+
+class ExtensionsElement:
```
done

libretto reviewed (commented):
@hackaugusto please see my comments... there are a few issues not finished yet.

libretto commented on karapace/protobuf/io.py:
```diff
@@ -0,0 +1,206 @@
+    return ".".join(result)
+
+
+def get_protobuf_class_instance(schema: ProtobufSchema, class_name: str, cfg: Dict) -> Any:
```
so there we will keep Any?

libretto commented on karapace/protobuf/kotlin_wrapper.py:
```diff
@@ -0,0 +1,59 @@
+    pass
+
+
+class KotlinRange:
```
done

amrutha-shanbhag commented:
Hi @hackaugusto , 

@libretto has addressed most of your feedback and left some comments of his own. Please let us know if there is any more feedback.

Thanks

hackaugusto reviewed (approved):
I added some comments, but those can be implemented as follow up PRs. Thank you for you contribution.

amrutha-shanbhag commented:
> I added some comments, but those can be implemented as follow up PRs. Thank you for you contribution.

thank you @hackaugusto  :)
